pandas/core/algorithms.py

```diff
@@ -253,84 +254,101 @@ def value_counts(values, sort=True, ascending=False, normalize=False,
-                is_datetimetz:
+    if bins is not None:
+        # TODO: This next line should be more efficient
+        result = result.reindex(np.arange(len(cat.categories)),
+                                fill_value=0)
+        result.index = bins[:-1]
 
-            if is_period:
-                values = PeriodIndex(values)
-            elif is_datetimetz:
-                tz = getattr(values, 'tz', None)
-                values = DatetimeIndex(values).tz_localize(None)
+    if sort:
+        result = result.sort_values(ascending=ascending)
 
-            values = values.view(np.int64)
-            keys, counts = htable.value_count_scalar64(values, dropna)
+    if normalize:
+        result = result / float(counts.sum())
 
-            if dropna:
-                msk = keys != iNaT
-                keys, counts = keys[msk], counts[msk]
+    return result
 
-            # localize to the original tz if necessary
-            if is_datetimetz:
-                keys = DatetimeIndex(keys).tz_localize(tz)
 
-            # convert the keys back to the dtype we came in
-            else:
-                keys = keys.astype(dtype)
+def _value_counts_arraylike(values, dropna=True):
+    is_datetimetz = com.is_datetimetz(values)
+    is_period = (isinstance(values, gt.ABCPeriodIndex) or
+                 com.is_period_arraylike(values))
 
-        elif com.is_integer_dtype(dtype):
-            values = com._ensure_int64(values)
-            keys, counts = htable.value_count_scalar64(values, dropna)
-        elif com.is_float_dtype(dtype):
-            values = com._ensure_float64(values)
-            keys, counts = htable.value_count_scalar64(values, dropna)
+    orig = values
 
-        else:
-            values = com._ensure_object(values)
-            mask = com.isnull(values)
-            keys, counts = htable.value_count_object(values, mask)
-            if not dropna and mask.any():
-                keys = np.insert(keys, 0, np.NaN)
-                counts = np.insert(counts, 0, mask.sum())
+    from pandas.core.series import Series
+    values = Series(values).values
+    dtype = values.dtype
 
-        if not isinstance(keys, Index):
-            keys = Index(keys)
-        result = Series(counts, index=keys, name=name)
+    if com.is_datetime_or_timedelta_dtype(dtype) or is_period:
+        from pandas.tseries.index import DatetimeIndex
+        from pandas.tseries.period import PeriodIndex
 
-        if bins is not None:
-            # TODO: This next line should be more efficient
-            result = result.reindex(np.arange(len(cat.categories)),
-                                    fill_value=0)
-            result.index = bins[:-1]
+        if is_period:
+            values = PeriodIndex(values)
+            freq = values.freq
 
-    if sort:
-        result = result.sort_values(ascending=ascending)
+        values = values.view(np.int64)
+        keys, counts = htable.value_count_scalar64(values, dropna)
 
```

Comment:
``sparse/categorical`` ``value_counts`` only supports ``dropna`` kwds. It may better to be an internal method.

If we do the same thing in ``DatetimeIndexOpsMixin``, the impl should be:
- ``DatetimeIndexOpsMixin`` 's public ``value_counts`` calls ``algorithms.py`` (as it needs to handle kwds like ``normalize``)
- ``algorithms.py`` convert datetime-likes to datetime-like ``Index`` to handle ``Series``
- Calls ``DatetimeIndexOpsMixin`` 's internal ``_value_counts``

Suppose to be little complex, or do you have anything in your mind?